tests/http/conftest.py

```diff
@@ -1,58 +1,54 @@
-import sys
-from typing import Type
+import importlib
+from typing import Any, Generator, Type
 
 import pytest
 
-from .clients import (
-    AioHttpClient,
-    AsgiHttpClient,
-    AsyncDjangoHttpClient,
-    AsyncFlaskHttpClient,
-    ChaliceHttpClient,
-    ChannelsHttpClient,
-    DjangoHttpClient,
-    FastAPIHttpClient,
-    FlaskHttpClient,
-    HttpClient,
-    SanicHttpClient,
-    StarliteHttpClient,
-    SyncChannelsHttpClient,
-)
-
-
-@pytest.fixture(
-    params=[
-        pytest.param(AioHttpClient, marks=pytest.mark.aiohttp),
-        pytest.param(AsgiHttpClient, marks=pytest.mark.asgi),
-        pytest.param(AsyncDjangoHttpClient, marks=pytest.mark.django),
-        pytest.param(AsyncFlaskHttpClient, marks=pytest.mark.flask),
-        pytest.param(ChaliceHttpClient, marks=pytest.mark.chalice),
-        pytest.param(DjangoHttpClient, marks=pytest.mark.django),
-        pytest.param(FastAPIHttpClient, marks=pytest.mark.fastapi),
-        pytest.param(FlaskHttpClient, marks=pytest.mark.flask),
-        pytest.param(SanicHttpClient, marks=pytest.mark.sanic),
-        pytest.param(ChannelsHttpClient, marks=pytest.mark.channels),
-        pytest.param(
-            # SyncChannelsHttpClient uses @database_sync_to_async and therefore
-            # needs pytest.mark.django_db
-            SyncChannelsHttpClient,
-            marks=[pytest.mark.channels, pytest.mark.django_db],
+from .clients.base import HttpClient
+
+
+def _get_http_client_classes() -> Generator[Any, None, None]:
+    for client, module, marks in [
```

Comment:
As a nitpick (Open Closed Principle) I'd move the loop outside of this function and iteratively call the body of the loop to within a generator comprehension in the params argument to the fixture

Reply:
can you show me how you'd do it? 😊

Reply:
(I'm begin lazy :P)

Reply:
Hey sorry Patrick I didn't have access to my PC when writing the comment, I just saw your message now. I'll post how to do it here when I get home 😊

Reply:
thanks! I forgot to ping you!

Reply:
So something like this:

```python
import importlib
from typing import Any, Sequence, Type

import pytest

from .clients.base import HttpClient


def _get_http_client_class(
    client: str,
    module: str,
    marks: Sequence[pytest.MarkDecorator],
) -> Any:
    try:
        client_class = getattr(
            importlib.import_module(f".{module}", package="tests.http.clients"),
            client,
        )
    except ImportError:
        client_class = None

    return pytest.param(
        client_class,
        marks=[
            *marks,
            pytest.mark.skipif(
                client_class is None, reason=f"Client {client} not found"
            ),
        ],
    )


@pytest.fixture(
    params=(
        _get_http_client_class(client, module, marks)
        for client, module, marks in [
            ("AioHttpClient", "aiohttp", [pytest.mark.aiohttp]),
            ("AsgiHttpClient", "asgi", [pytest.mark.asgi]),
            ("AsyncDjangoHttpClient", "async_django", [pytest.mark.django]),
            ("AsyncFlaskHttpClient", "async_flask", [pytest.mark.flask]),
            ("ChannelsHttpClient", "channels", [pytest.mark.channels]),
            ("ChaliceHttpClient", "chalice", [pytest.mark.chalice]),
            ("DjangoHttpClient", "django", [pytest.mark.django]),
            ("FastAPIHttpClient", "fastapi", [pytest.mark.fastapi]),
            ("FlaskHttpClient", "flask", [pytest.mark.flask]),
            ("SanicHttpClient", "sanic", [pytest.mark.sanic]),
            ("StarliteHttpClient", "starlite", [pytest.mark.starlite]),
            (
                "SyncChannelsHttpClient",
                "channels",
                [pytest.mark.channels, pytest.mark.django_db],
            ),
        ]
    )
)
def http_client_class(request: Any) -> Type[HttpClient]:
    return request.param


@pytest.fixture()
def http_client(http_client_class: Type[HttpClient]) -> HttpClient:
    return http_client_class()
```

Or alternatively:

```python
import importlib
from typing import Any, Generator, Sequence, Tuple, Type

import pytest

from .clients.base import HttpClient


def parametrize_clients(
    specifications: Sequence[Tuple[str, str, Sequence[pytest.MarkDecorator]]],
) -> Generator[Any, None, None]:
    for client, module, marks in specifications:
        try:
            client_class = getattr(
                importlib.import_module(f".{module}", package="tests.http.clients"),
                client,
            )
        except ImportError:
            client_class = None

        yield pytest.param(
            client_class,
            marks=[
                *marks,
                pytest.mark.skipif(
                    client_class is None, reason=f"Client {client} not found"
                ),
            ],
        )


@pytest.fixture(
    params=parametrize_clients(
        [
            ("AioHttpClient", "aiohttp", [pytest.mark.aiohttp]),
            ("AsgiHttpClient", "asgi", [pytest.mark.asgi]),
            ("AsyncDjangoHttpClient", "async_django", [pytest.mark.django]),
            ("AsyncFlaskHttpClient", "async_flask", [pytest.mark.flask]),
            ("ChannelsHttpClient", "channels", [pytest.mark.channels]),
            ("ChaliceHttpClient", "chalice", [pytest.mark.chalice]),
            ("DjangoHttpClient", "django", [pytest.mark.django]),
            ("FastAPIHttpClient", "fastapi", [pytest.mark.fastapi]),
            ("FlaskHttpClient", "flask", [pytest.mark.flask]),
            ("SanicHttpClient", "sanic", [pytest.mark.sanic]),
            ("StarliteHttpClient", "starlite", [pytest.mark.starlite]),
            (
                "SyncChannelsHttpClient",
                "channels",
                [pytest.mark.channels, pytest.mark.django_db],
            ),
        ]
    )
)
def http_client_class(request: Any) -> Type[HttpClient]:
    return request.param


@pytest.fixture()
def http_client(http_client_class: Type[HttpClient]) -> HttpClient:
    return http_client_class()

```

Could then re-use the same `parametrize_clients` in the websockets conftest